### ISSUE TITLE
s3 object endpoints only work with AWS

### ISSUE THREAD

**Issue by jeffvance** [kubevirt/containerized-data-importer#389]
Now that we're creating full lifecycle functional tests we need to automate the testing of s3. `s3` is an amazon protocol and is not supported by minio (our object store client). Instead, minio provides object store over http (as does amazon). As far as we know, we cannot test object store endpoints using the s3 protocol unless the file lives on aws.

Before formal functional tests, we manually tested s3 using an amazon endpoint such as `s3://<bucket>/<objectname>` (eg: _s3://kubevirt-images/foo.iso_), or, over http, such as `https://s3.amazonaws.com/<bucket>/<objectname>` (eg. _https://s3.amazonaws.com/kubevirt-images/foo.iso.tar`_).

In short, we cannot functionally _test_ the importing of s3 objects, in our current test env, unless the file under test resides on aws s3.

Possible solutions:
1. remove cdi support of the s3 scheme and only support non-object storage. 
1. add a new object store endpoint annotation, eg. `cdi.kubevirt.io/obj-storage.import.endpoint=`, where **obj** is new. Keep the existing s3 code and call the s3 reader function when this anno is present or when the url scheme is `s3`.
1. ?

**Comment by jeffvance** [comment]
@copejon  @davidvossel  @aglitke  @nellyc

**Comment by copejon** [comment]
3...? - Drop the existing endpoint annotation and replace it with 2 distinct ones (file and object).  

Minio clients can process both s3:// and http://, which is great.  Where this has thrown a wrench into things is that we can't distinguish between an object store and a remote file by the URL alone when using http.  The earliest CDI design favored a simple api over this differentiation.  This led us to only supporting s3://, amazon's proprietary url scheme.

By disambiguating the storage types, we can cleanly process both and support all the other object store solutions.

Bonus - it'll allow me to finish #376 for the s3 protocol.

**Comment by jeffvance** [comment]
> Drop the existing endpoint annotation and replace it with 2 distinct ones (file and object).

That's what I was trying to convey in choice 2 above but Jon stated it better What's different is that Jon implies (perhaps) that we'd create two new endpoint names with the distinguishing feature being file vs. object storage. Maybe something like `cdi.kubevirt.io/obj-storage.import.endpoint=` and `cdi.kubevirt.io/file-storage.import.endpoint=`

**Comment by davidvossel** [comment]
> we can't distinguish between an object store and a remote file by the URL alone when using http

why would we need to do this? In either case aren't we just downloading data from a url?

> As far as we know, we cannot test object store endpoints using the s3 protocol unless the file lives on aws.

I thought that's what minio is designed to do is it not?  The minio server is capable of supporting s3 protocol v4.

**Comment by copejon** [comment]
> why would we need to do this? In either case aren't we just downloading data from a url?

The issue is in how we communicate with the endpoint.  With an object store, auth requires that we use an s3 client.  So this issue has 2 components - 

1 - how do we know which client to use (http / s3)
2 - fix the hardcoded `s3.amazonaws.com` endpoint for [`s3://` protocols](url).  this would allow us to hit object stores with http endpoints (which is every one of them -aws).  This is reliant on how we solve 1

**Comment by davidvossel** [comment]
> To clarify - s3://bucket/file resolves to https://s3.amazonaws.com/bucket/file. By queuing off of the s3:// url scheme in the importer, we lock ourselves into connections only to aws.

ah, okay I see what you're getting at now. 

>  Maybe something like cdi.kubevirt.io/storage.import.obj-endpoint= and cdi.kubevirt.io/storage.import.file-endpoint=.

I'm not sure 'obj' vs 'file' is granular enough to express all the different endpoint combinations. Object storage isn't necessarily just s3.  This change would also be backwards incompatible unless we kept the old annotations as well.

What about this. We add a 'type' annotation that lets users explicitly defines how the client should import the data.

```
cdi.kubevirt.io/obj-storage.import.type
```

type=s3 means the s3 client should be used. Type=http means a plain http get should be used. There may be other types implemented in the future. 

This change would be backwards compatible with what we have now. If someone sets  ```cdi.kubevirt.io/obj-storage.import.endpoint=s3://bucket/file```, then we default the type to s3 if type wasn't defined. Likewise, if someone sets type=s3 and then the endpoint is an http:// address, we still use the s3 client.

Would that work?

**Comment by jeffvance** [comment]
I like that suggestion @davidvossel . It solves this problem, is extensible, and provides backwards compatibility. Nice!

**Comment by copejon** [comment]
@davidvossel Sounds good to me!
